apps/emqx_management/src/emqx_mgmt_api_clients.erl

```diff
@@ -1165,6 +1273,32 @@ format_persistent_session_info(ClientId, PSInfo0) ->
     ),
     result_format_undefined_to_null(PSInfo).
 
+format_msg(#message{
+    id = ID,
+    qos = Qos,
+    topic = Topic,
+    from = From,
+    timestamp = Timestamp,
+    headers = Headers,
+    payload = Payload
+}) ->
+    Msg = #{
+        msgid => emqx_guid:to_hexstr(ID),
+        qos => Qos,
+        topic => Topic,
+        publish_at =>
+            emqx_utils_calendar:epoch_to_rfc3339(Timestamp),
+        from_clientid => emqx_utils_conv:bin(From),
+        from_username => maps:get(username, Headers, <<>>)
+    },
+    case erlang:byte_size(Payload) =< ?MAX_MSG_PAYLOAD_SIZE of
```

Comment:
NOTE: message formatter/schema is copied from emqx_retainer_api

Reply:
Maybe add an option like `?payload=none | base64  | plain` to the API.
Return 400 if payload is not JSON serializable and `plain` is requested.

About size limit, maybe alongside the `page` limit, there can be a `max_payload_bytes` limit and each response try to accumulate not greater than this limit however ensure at least one message is returned even if the size is above limit.

Reply:
Done